// SPDX-FileCopyrightText: 2025 Carnegie Mellon University - LivingEdgeLab
//
// SPDX-License-Identifier: GPL-2.0-only

syntax = "proto3";
package protocol.mission_service;

import "common.proto"

/*
 * Used to start a new mission or stop an active mission
 */
service MissionService {
  // Upload a mission for execution
  rpc Upload (UploadRequest) returns (UploadResponse) {}
  // Start an uploaded mission
  rpc Start (StartRequest) returns (StartResponse) {}
  // Stop the current mission
  rpc Stop (StopRequest) returns (StopResponse) {}
  // Send a notification to the current mission
  rpc Notify (NotifyRequest) returns (NotifyResponse) {}
  // Get info on the current mission
  rpc ConfigureTelemetryStream (ConfigureTelemetryStreamRequest) returns (ConfigureTelemetryStreamResponse) {}
}

message MissionData {
  string uri = 1; // URI, either local or remote, of a mission file
}

message UploadRequest {
  Request request = 1;
  MissionData mission = 2; // Data of the target mission
}

message UploadResponse {
  Response response = 1;
}

message StartRequest {
  Request request = 1;
}

message StartResponse {
  Response response = 1;
}

message StopRequest {
  Request request = 1;
}

message StopResponse {
  Response response = 1;
}

message NotifyRequest {
  Request request = 1;
  int32 notify_code = 2; // Integer notification code, generated by the backend 
}

message NotifyResponse {
  Response response = 1;
}

message GetCurrentRequest {
  Request request = 1;
}

message GetCurrentResponse {
  Response response = 1;
  MissionData current = 2; // Data of the current mission
}
